govcd/nsxv_ipset.go

```diff
@@ -0,0 +1,254 @@
+		return nil, err
+	}
+
+	allIpSets, err := vdc.GetAllNsxvIpSets()
+	if err != nil {
+		return nil, err
+	}
+
+	util.Logger.Printf("[DEBUG] Searching for IP set with name: %s", name)
+	for _, ipSet := range allIpSets {
+		util.Logger.Printf("[DEBUG] Checking IP set: %#+v", ipSet)
+		if ipSet.Name != "" && ipSet.Name == name {
+			return ipSet, nil
+		}
+	}
+
+	return nil, ErrorEntityNotFound
+}
+
+// GetNsxvIpSetById searches for IP set by ID. Returns ErrorEntityNotFound if an IP set is not found
+func (vdc *Vdc) GetNsxvIpSetById(id string) (*types.EdgeIpSet, error) {
+	if err := validateGetNsxvIpSet(id, ""); err != nil {
+		return nil, err
+	}
+
+	allIpSets, err := vdc.GetAllNsxvIpSets()
+	if err != nil {
+		return nil, err
+	}
+
+	util.Logger.Printf("[DEBUG] Searching for IP set with id: %s", id)
+	for _, ipSet := range allIpSets {
+		util.Logger.Printf("[DEBUG] Checking IP set: %#+v", ipSet)
+		if ipSet.ID != "" && ipSet.ID == id {
+			return ipSet, nil
+		}
+	}
+
+	return nil, ErrorEntityNotFound
+}
+
+// GetNsxvIpSetByNameOrId uses the same identifier to search by name and by ID. Priority is to try
+// and find the IP set by ID. If it is not found - then a search by name is performed.
+func (vdc *Vdc) GetNsxvIpSetByNameOrId(identifier string) (*types.EdgeIpSet, error) {
+	getByName := func(name string, refresh bool) (interface{}, error) { return vdc.GetNsxvIpSetByName(name) }
+	getById := func(id string, refresh bool) (interface{}, error) { return vdc.GetNsxvIpSetById(id) }
+	entity, err := getEntityByNameOrId(getByName, getById, identifier, true)
+	if entity == nil {
+		return nil, err
+	}
+	return entity.(*types.EdgeIpSet), err
+}
+
+// GetAllNsxvIpSets retrieves all IP sets and returns []*types.EdgeIpSet or an
+// error of type ErrorEntityNotFound if there are no IP sets
+func (vdc *Vdc) GetAllNsxvIpSets() ([]*types.EdgeIpSet, error) {
+	vdcId, err := getUuidFromHref(vdc.Vdc.HREF)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get vdc ID from HREF: %s", err)
+	}
+
+	// build a path for to read all IP sets in a scope. A scope is defined by vDC ID. The endpoint
+	// should look like:
+	// https://192.168.1.109/network/services/ipset/scope/f9daf2da-b4f9-4921-a2f4-d77a943a381c where
+	// the trailing UUID is vDC ID
+	httpPath, err := vdc.buildNsxvNetworkServiceEndpointURL(types.NsxvIpSetServicePath + "/scope/" + vdcId)
+	if err != nil {
+		return nil, fmt.Errorf("could not get network services API endpoint for IP set: %s", err)
+	}
+
+	// Anonymous struct to unwrap list of IP sets <list><ipset></ipset><ipset></ipset></list>
+	ipSetsResponse := &struct {
+		XMLName          xml.Name `xml:"list"`
+		types.EdgeIpSets `xml:"ipset"`
+	}{}
+
+	// This query returns all IP sets on the scope (scoped by vDC ID)
+	errString := fmt.Sprintf("unable to read IP sets for scope %s: %%s", vdcId)
+	_, err = vdc.client.ExecuteRequest(httpPath, http.MethodGet, types.AnyXMLMime, errString, nil, ipSetsResponse)
+	if err != nil {
+		return nil, err
```

Comment:
Custom message on what action error happen would be helpfull

Reply:
The error message is 2 lines above in line 168